src/box/lua/init.c

```diff
@@ -590,29 +590,33 @@ lbox_txn_pairs(struct lua_State *L)
 {
 	int64_t txn_id = luaL_toint64(L, lua_upvalueindex(1));
 	struct txn *txn = in_txn();
-	if (txn == NULL || txn->id != txn_id) {
+	struct txn_stmt *stmt;
+	stmt = (struct txn_stmt *)lua_topointer(L, lua_upvalueindex(2));
+	if (txn == NULL || (stmt != NULL && txn->id != txn_id)) {
 		diag_set(ClientError, ER_CURSOR_NO_TRANSACTION);
 		return luaT_error(L);
 	}
 	luaL_pushint64(L, txn_id);
-	lua_pushlightuserdata(L, stailq_first_entry(&txn->stmts,
-						    struct txn_stmt, next));
+	lua_pushlightuserdata(L, stmt);
 	lua_pushcclosure(L, lbox_txn_iterator_next, 2);
 	lua_pushnil(L);
 	lua_pushinteger(L, 0);
 	return 3;
 }
 
 /**
- * Push an argument for on_commit Lua trigger. The argument is
+ * Push an argument for on_commit and on_rollback Lua triggers. The argument is
  * a function to open an iterator over the transaction statements.
  */
 static int
 lbox_push_txn(struct lua_State *L, void *event)
 {
-	struct txn *txn = (struct txn *) event;
-	luaL_pushint64(L, txn->id);
-	lua_pushcclosure(L, lbox_txn_pairs, 1);
+	struct txn_stmt *stmt = (struct txn_stmt *)event;
+	/* stmt is NULL for empty transactions. */
+	int64_t txn_id = stmt != NULL ? stmt->txn->id : -1;
```

Comment:
Since `stmt` can be `NULL`, wouldn't it be better to use `in_txn()` here and in other triggers instead of `stmt->txn`?

Reply:
Done.